### DISCUSSION TITLE
Stopping endless loops clean on shutdown

### DISCUSSION THREAD

**Discussion by fliiiix** [fastapi/fastapi#9373] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

import asyncio

app = FastAPI()
running = True

@app.on_event("shutdown")
def shutdown_event():
    global running
    running = False

@app.get("/")
async def index():
    while running:
        await asyncio.sleep(0.1)
```

### Description

- run it as uvicorn module.filename:app --host 0.0.0.0
- curl http://ip:port/
- then stop the server (pressing CTRL+C)
- observe that the server is not stopped and `shutdown_event` is never called

### Operating System

Linux

### FastAPI Version

0.95.0

### Python Version

3.10.10

### Additional Context

There is a https://stackoverflow.com/questions/75975807/how-to-stop-a-loop-on-shutdown-in-fastapi/75978619#75978619 for this

**Comment by Kludex** [comment]
This works as expected.

The shutdown event only runs after all connections are closed. There's a feature request on uvicorn to add `--timeout-shutdown` to mitigate this "issue".

**Reply by fliiiix** [reply]
Then I would argue the docs are wrong because they state:
> There's an alternative way to define this logic to be executed during startup and during shutdown.
https://fastapi.tiangolo.com/advanced/events/#alternative-events-deprecated

And to me it's still feels like there is something missing where you get informed that a shutdown is requested so the app has a chance to stop long running things

**Reply by Kludex** [reply]
Why is it wrong, exactly?

**Reply by fliiiix** [reply]
for me it is missing the ->  The shutdown event only runs after all connections are closed. part.
At least to me in contrast to the new way it sounded like `@app.on_event("shutdown")` is started sometime during shutdown without waiting for all the requests to finish.

**Reply by Kludex** [reply]
I mean, what in the documentation looks wrong considering my statement?

**Reply by fliiiix** [reply]
I don't think its wrong but it is missing the part that shutdown happens after all requests are finished

So maybe 

> There's an alternative way to define this logic to be executed during startup and during shutdown.
**The same rules as for a `lifespan` apply.**


would be enough

**Reply by fliiiix** [reply]
@Kludex what do you think about updating the wording?

**Comment by 8thgencore** [comment, marked as answer]
```python3
import signal

app = FastAPI()
running = True

def stop_server(*args):
    global running
    running = False

@app.on_event("startup")
def startup_event():
    signal.signal(signal.SIGINT, stop_server)

@app.get("/")
async def index():
    while running:
        await asyncio.sleep(0.1)
```

With these changes, when you stop the server using , the function will be called, and the infinite loop in the  function will exit, allowing the server to shut down cleanly

**Reply by Kludex** [reply]
The server doesn't shut down cleanly. It shutdowns immediately, which is not the same...

All background tasks and other requests are not given time to finish.

**Reply by fliiiix** [reply]
Why not? It seems like `signal.SIGINT` is sent to the server regardless now we just use it as a hock to stop our loop which finishes the request on `/` and all the rest should be the same?

**Reply by Kludex** [reply]
Right, sorry. It does work... :thinking:

**Reply by fliiiix** [reply]
i mean as long as you don't send a `sys.exit` yourself it should be fine, but it also relies on your code that it ends after setting a flag to shutdown which needs to be cooperative (but i can control that in my case)

**Reply by Jakobhenningjensen** [reply]
This seems like it is caught *after* the lifespan-shutdown event btw.
